app/models/papi_client.rb

```diff
@@ -0,0 +1,307 @@
+    Google::Apis::GenomicsV2alpha1::Action.new(
+        commands: commands,
+        environment: environment,
+        flags: flags,
+        image_uri: INGEST_DOCKER_IMAGE,
+        labels: labels,
+        timeout: timeout
+    )
+  end
+
+  # Instatiate a resources object to tell where to run a pipeline
+  #
+  # * *params*
+  #   - regions: (Array<String>) => An array of GCP regions allowed for VM allocation
+  #
+  # * *return*
+  #   - (Google::Apis::GenomicsV2alpha1::Resources)
+  def create_resources_object(regions:)
+    Google::Apis::GenomicsV2alpha1::Resources.new(
+         project_id: COMPUTE_PROJECT,
+         regions: regions,
+         virtual_machine: self.create_virtual_machine_object
+    )
+  end
+
+  # Instantiate a VM object to specify in resources.  Assigns the portal service account to the VM
+  # to manage permissions
+  #
+  # * *params*
+  #   - +machine_type+ (String) => GCP VM machine type (defaults to 'n1-standard-1')
+  #   - +preemptible+ (Boolean) => Indication of whether VM can be preempted (defaults to false)
+  # * *return*
+  #   - (Google::Apis::GenomicsV2alpha1::VirtualMachine)
+  def create_virtual_machine_object(machine_type: 'n1-standard-1', preemptible: false)
+    Google::Apis::GenomicsV2alpha1::VirtualMachine.new(
+        machine_type: machine_type,
+        preemptible: preemptible,
+        service_account: Google::Apis::GenomicsV2alpha1::ServiceAccount.new(email: self.issuer, scopes: GOOGLE_SCOPES)
+    )
+  end
+
+  # Determine command line to pass to ingest based off of file & action requested
+  #
+  # * *params*
+  #   - +study_file+ (StudyFile) => StudyFile to be ingested
+  #   - +action+ (String/Symbol) => Action to perform on ingest
+  #
+  # * *return*
+  #   - (Array) Command Line, in Docker "exec" format
+  #
+  # * *raises*
+  #   - (ArgumentError) => The requested StudyFile and action do not correspond with each other, or cannot be run yet
+  def get_command_line(study_file:, action:)
+    validate_action_by_file(action, study_file)
+    study = study_file.study
+    command_line = "python ingest_pipeline.py --study-accession #{study.accession} --file-id #{study_file.id} #{action}"
+    case action.to_s
+    when 'ingest_expression'
+      if study_file.file_type == 'Expression Matrix'
+        command_line += " --matrix-file #{study_file.gs_url} --matrix-file-type dense"
+      elsif study_file.file_type === 'MM Coordinate Matrix'
+        bundled_files = study_file.bundled_files
+        genes_file = bundled_files.detect {|f| f.file_type == '10X Genes File'}
+        barcodes_file = bundled_files.detect {|f| f.file_type == '10X Barcodes File'}
+        command_line += " --matrix-file #{study_file.gs_url} --matrix-file-type mtx" \
+                      " --gene-file #{genes_file.gs_url} --barcode-file #{barcodes_file.gs_url}"
+      end
+    when 'ingest_cell_metadata'
+      command_line += " --cell-metadata-file #{study_file.gs_url} --ingest-cell-metadata"
```

Comment:
It doesn't concern this PR per se, but I'm still unsure why `--ingest-cell-metadata` is needed.  It seems like it should already be entailed by the subparser argument `ingest_cell_metadata`.

(And similarly for `ingest_cluster` / `--ingest-cluster`, and `ingest_subsample` / `--subsample`.)

Reply:
This has to deal with how we tell ingest what to ingest. Sub-parsers are not captured in arguments. Cluster, cell metadata, and subsampling sub-parsers share argument names that are the same. We simply cannot check to see if an argument exist to see what function to invoke.  If you don't want --ingest-cell-metdata we should think of other ways to invoke ingest of cell-metadata, cluster and subsampling.

Reply:
> Sub-parsers are not captured as arguments from argparse

That's the piece I was missing.  Thanks for helping me understand!

Reply:
Yeah, I have the same concern, and while I think that there _might_ be a better way of doing this, it does make sense in terms of the hierarchy of the parsers/parameters.